modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java

```diff
@@ -5168,6 +5253,138 @@ public CacheConfigurationEnricher enricher() {
         return enricher;
     }
 
+
+    /** {@inheritDoc} */
+    @Override public void onReadyForRead(ReadOnlyMetastorage metastorage) {
+        synchronized (metaStorageMux) {
+            if (localContinuousTasks.isEmpty()) {
+                try {
+                    metastorage.iterate(
+                        STORE_LOCAL_CONTINUOUS_TASK_PREFIX,
+                        (key, val) -> localContinuousTasks.put(key, (LocalContinuousTask)val),
+                        true
+                    );
+                }
+                catch (IgniteCheckedException e) {
+                    throw new IgniteException("Failed to iterate continuous tasks storage.", e);
+                }
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onReadyForReadWrite(ReadWriteMetastorage metastorage) {
+        synchronized (metaStorageMux) {
+            try {
+                for (Map.Entry<String, LocalContinuousTask> entry : localContinuousTasks.entrySet()) {
+                    if (metastorage.readRaw(entry.getKey()) == null)
+                        metastorage.write(entry.getKey(), entry.getValue());
+                }
+            }
+            catch (IgniteCheckedException e) {
+                throw new IgniteException("Failed to read key from continuous tasks storage.", e);
+            }
+        }
+
+        this.metastorage = metastorage;
+    }
+
+    /**
+     * Builds a metastorage key for continuous task object.
+     *
+     * @param obj Object.
+     * @return Metastorage key.
+     */
+    private String localContinuousTaskMetastorageKey(LocalContinuousTask obj) {
+        String k = STORE_LOCAL_CONTINUOUS_TASK_PREFIX + obj.shortName();
+
+        if (k.length() > MetastorageTree.MAX_KEY_LEN) {
+            int hashLenLimit = 5;
+
+            String hash = String.valueOf(k.hashCode());
+
+            k = k.substring(0, MetastorageTree.MAX_KEY_LEN - hashLenLimit) +
+                (hash.length() > hashLenLimit ? hash.substring(0, hashLenLimit) : hash);
+        }
+
+        return k;
+    }
+
+    /**
+     * Adds local continuous task object.
+     *
+     * @param obj Object.
+     */
+    public void addLocalContinuousTask(LocalContinuousTask obj) {
+        String objName = localContinuousTaskMetastorageKey(obj);
+
+        synchronized (metaStorageMux) {
+            localContinuousTasks.put(objName, obj);
+
+            if (metastorage != null) {
+                ctx.cache().context().database().checkpointReadLock();
```

Comment:
why do we need rlock here ?

Reply:
Metastorage write/remove operations modify page memory, so we need read lock here.